Remove host container migration

vyaghras opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #

**Description of changes:**
```
migrations: remove all the weak setting on downgrade 
migrations: Change public control host-container source metadata to object 
migrations: Change aws admin host-container source metadata to object 
migrations: Change public control host-container source metadata to object 
migrations: Change public admin host-container source metadata to object 
common_migrations: add new migrations to handle settings-generator as struct 
models: Add Strength enum and Setting-Generator struct 
apiclient: update README to document version 2 of /tx API 
shared-defaults: change public host container source setting-generator to table 
shared-defaults: change aws host container source setting-generator to table 
datastore: changes to match changes in Core kit datastore
```


**Testing done:**
Refer to testing in https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/294


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

vyaghras committed:
datastore: changes in Core kit datastore

Updated the `commit_transaction` function to enable committing
metadata from pending transactions. In commit transaction we will first
commit metadata and then pending keys to correctly perform the check to
identify if key exists or not.

The strength handling among pending and committed transaction is as:
If pending metadata is strong and committed metadata is weak, commit
the pending setting.

If pending metadata is weak, committed metadata is strong and
the setting is already available, do not downgrade from strong to
weak. Otherwise add the strength file with value as weak.

If pending and committed metadata are the same, no action is
performed.

Additionally, made minor changes to metadata functions for improved
access and flexibility:
- Introduced a `committed` field to dynamically access metadata in
 pending transactions.
- Replaced the hardcoded use of live committed metadata with this
 committed variable ans pass Committed::Live from previous usages.

Refer commit:
https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/294/commits/20a435e3f06625f6c75ccf8109a4a722a055d229
Refer PR:
https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/294

vyaghras committed:
common_migrations: add new migrations to handle settings-generator as struct

- RemoveMetadataAndWeakSettingsMigration: When we downgrade multiple
  version to a version where migrator is not aware of deleting the
  setting-generator as struct or the strength file. This migration
  will help us to delete the weak settings and all metadata.

- RemoveSchnauzerMigration: This will remove metadata(always) and data
  if the value matches with the value in existing datastore.

- RemoveMatchingString: This will remove the data when value is same
  as the old value for public host containers.

vyaghras committed:
shared-defaults: change aws host container source setting-generator to table

Update the source from string like:
schnauzer-v2 render --requires 'aws@v1(helpers=[ecr-prefix])' --template
'{{ ecr-prefix settings.aws.region }}/bottlerocket-admin:v0.11.14'

To setting generator as object, that contains:
- command
- strength
- skip-if-populated

vyaghras committed:
shared-defaults: add strength metadata for public host container

Add a setting-generator that will populate the host-containers
source and add strength=weak metadata if the setting is not
populated.

vyaghras committed:
models: Add Strength enum and Setting-Generator struct

We need to add these to match changes in Bottlerocket-core-kit
Refer commit:
https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/294/commits/a72f6bd16f7de0d628e80fb03bbf827ebd4892e2

PR: https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/294

vyaghras committed:
apiclient: update README to document version 2 of /tx API

vyaghras commented:
☝️ Changed the Version in Twoliter.toml

bcressey commented on sources/shared-defaults/aws-host-containers.toml:
```diff
@@ -12,5 +14,8 @@ setting-generator = "shibaken generate-admin-userdata"
+command = "schnauzer-v2 render --requires 'aws@v1(helpers=[ecr-prefix])' --template '{{ ecr-prefix settings.aws.region }}/bottlerocket-control:v0.7.18'"
+strength = "weak"
+skip-if-populated = true
+
```
```suggestion
```

bcressey commented on sources/shared-defaults/aws-host-containers.toml:
```diff
@@ -2,8 +2,10 @@
+[metadata.settings.host-containers.admin.source.setting-generator]
+command = "schnauzer-v2 render --requires 'aws@v1(helpers=[ecr-prefix])' --template '{{ ecr-prefix settings.aws.region }}/bottlerocket-admin:v0.11.14'"
+strength = "weak"
+skip-if-populated = true
```
Shouldn't `skip-if-populated` default to true? Can we omit this?

bcressey commented on sources/api/datastore/src/error.rs:
```diff
@@ -59,6 +59,12 @@ pub enum Error {
+    Serialize { source: serde_json::Error },
+
+    #[snafu(display("Unable to de-serialize data: {}", source))]
+    DeSerialize { source: serde_json::Error },
```
```suggestion
    #[snafu(display("Unable to deserialize data: {}", source))]
    Deserialize { source: serde_json::Error },
```

bcressey commented on sources/settings-migrations/v1.29.0/change-public-control-container-to-set-gen/src/main.rs:
```diff
@@ -0,0 +1,27 @@
+        setting: "settings.host-containers.control.source",
+        old_val: OLD_CONTROL_CTR,
+        new_val: NEW_CONTROL_CTR,
+        setting_gen_command: "emitter public.ecr.aws/bottlerocket/bottlerocket-control:v0.7.17",
```
I don't understand where `emitter` comes from.

bcressey commented on sources/settings-migrations/v1.29.0/change-public-control-container-to-set-gen/src/main.rs:
```diff
@@ -0,0 +1,27 @@
+    migrate(ReplaceSettingWithSettingGeneratorMigration {
+        setting: "settings.host-containers.control.source",
+        old_val: OLD_CONTROL_CTR,
+        new_val: NEW_CONTROL_CTR,
```
It's confusing to me to mix settings changes with metadata changes.

I would model this with a pair of migration types:
* `RemoveSchnauzerMigration` - like `ReplaceSchnauzerMigration`, but removes the setting on upgrade if it matches. No-op on downgrade, like `RemoveMetadataMigration`.
* Maybe `ResetMetadataMigration` where the upgrade and downgrade action is the same, and we delete the metadata both ways, so that `storewolf` fixes it up? In effect this would combine the behaviors of `AddMetadataMigration` and `RemoveMetadataMigration`.

Edit: having seen the `RemoveWeakSettingsMigration` - I think you want a similar `ResetMetadataMigration` that clears out all metadata for all settings. Otherwise it'll be easy to lose track of which settings might have metadata that persists on downgrade.

bcressey commented on sources/api/migration/migration-helpers/src/common_migrations.rs:
```diff
@@ -1914,3 +2259,56 @@ impl Migration for NoOpMigration {
+        Ok(input)
+    }
+
+    /// No work to do on backward migrations, copy the same datastore
```
Comment seems incorrect?

bcressey commented on sources/api/migration/migration-helpers/src/common_migrations.rs:
```diff
@@ -1914,3 +2259,56 @@ impl Migration for NoOpMigration {
+                inner_map.remove("strength");
+                input.metadata.insert(key.clone(), inner_map);
+            }
+            input.data.remove(&key);
```
nit: add a comment that we're also removing the setting, not just the "strength" metadata

bcressey commented on sources/api/migration/migration-helpers/src/common_migrations.rs:
```diff
@@ -1914,3 +2259,56 @@ impl Migration for NoOpMigration {
+                inner_map.remove("setting-generator");
+                input.metadata.insert(key.clone(), inner_map);
+            }
+        }
```
I don't think we should touch settings generators here
```suggestion
```

cbgbt commented on sources/shared-defaults/public-host-containers.toml:
```diff
@@ -6,9 +6,17 @@
+[metadata.settings.host-containers.admin.source.setting-generator]
+command = "schnauzer-v2 render --template 'public.ecr.aws/bottlerocket/bottlerocket-admin:v0.11.14'"
+strength = "weak"
+skip-if-populated = true
```
Since these are static strings and don't require templates, can we disregard the setting generator and accomplish the same thing like this?

```suggestion
[settings.host-containers.admin]
enabled = false
superpowered = true
source = "public.ecr.aws/bottlerocket/bottlerocket-admin:v0.11.14"

[metadata.settings.host-containers.admin.source]
strength = "weak"
```

cbgbt commented on sources/shared-defaults/public-host-containers.toml:
```diff
@@ -6,9 +6,17 @@
+[metadata.settings.host-containers.control.source.setting-generator]
+command = "schnauzer-v2 render --template 'public.ecr.aws/bottlerocket/bottlerocket-control:v0.7.18'"
+strength = "weak"
+skip-if-populated = true
```
Same feedback as the `admin` container source for public ECR.

cbgbt commented on sources/api/migration/migration-helpers/src/common_migrations.rs:
```diff
@@ -1157,6 +1262,246 @@ mod test_replace_schnauzer_migration {
 // =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
+// We use this migration to change the aws host-containers setting-generator to struct from string.
+// A metadata strength as `weak`, shows that associated setting is a weak setting.
```
nit: use docstring comments

Less nit: A brief amount of context on string vs struct is useful here, since we're adding this to `common_migrations` and theoretically someone will need to read this comment and code again in the future.

e.g.

```suggestion
/// We use this migration to change the aws host-containers setting-generator to struct from string.
/// A metadata strength as `weak`, shows that associated setting is a weak setting.
///
/// String `setting-generators` only provide the command to run, using the default
/// values for `strength` and `skip-if-populated` ("weak", and "true").
/// Struct generators allow customization of those fields.
```

cbgbt commented on sources/api/migration/migration-helpers/src/common_migrations.rs:
```diff
@@ -1157,6 +1262,246 @@ mod test_replace_schnauzer_migration {
+    pub metadata_type: &'static str,
+    pub binary_for_generator: &'static str,
+    pub skip_if_populated: bool,
+}
```
This seems highly specific to `setting-generators`, I doubt we'd use it elsewhere. I'd suggest renaming it to be more specific and dropping `metadata_type` as an argument.

We're already using `Shlex` ("shell lexer") in this migrator. Do we need `binary_for_generator` if we can just parse it out? The binary should always be the first element of the lexed command line.

Finally, I wonder if the settings for the "new" generator should just be a `SettingsGenerator` struct?

vyaghras committed:
Upgrade Bottlerocket version to 1.33.0

bcressey commented on sources/settings-migrations/v1.33.0/reset-metadata-migration/src/main.rs:
```diff
@@ -0,0 +1,18 @@
+use migration_helpers::{migrate, Result};
+use std::process;
+
+// Remove the weak settings on downgrade
+fn run() -> Result<()> {
+    migrate(ResetMetadataMigration)
+}
```
nit: `ResetMetadataMigration` could be embedded here since it's unlikely we'll make use of it as a common migration

vyaghras replied:
Replaced this migration with RemoveMetadataAndWeakSettingsMigration.

cbgbt replied:
I think this feedback still applies -- `RemoveMetadataAndWeakSettingsMigration` shouldn't need to live in `common_migrations`, since we shouldn't really need to run it beyond the `1.32` -> `1.33` upgrade.

bcressey commented on sources/api/migration/migration-helpers/src/common_migrations.rs:
```diff
@@ -1912,3 +1788,250 @@ impl Migration for NoOpMigration {
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// We use this migration to remove a setting string if it matches the old value.
+pub struct RemoveOldData {
+    pub setting: &'static str,
+    pub old_val: &'static str,
+    pub new_val: &'static str,
+}
```
`new_val` not used. Maybe rename this to `RemoveMatchingString`

bcressey commented on sources/settings-migrations/v1.33.0/public-remove-source-control/src/main.rs:
```diff
@@ -0,0 +1,25 @@
+use migration_helpers::{migrate, Result};
+use std::process;
+
+const OLD_CONTROL_CTR: &str = "public.ecr.aws/bottlerocket/bottlerocket-control:v0.7.18";
+const NEW_CONTROL_CTR: &str = "public.ecr.aws/bottlerocket/bottlerocket-control:v0.7.18";
```
Needs to be the latest public control container

bcressey commented on sources/settings-migrations/v1.33.0/public-remove-source-admin/src/main.rs:
```diff
@@ -0,0 +1,25 @@
+use migration_helpers::{migrate, Result};
+use std::process;
+
+const OLD_ADMIN_CTR: &str = "public.ecr.aws/bottlerocket/bottlerocket-admin:v0.11.14";
+const NEW_ADMIN_CTR: &str = "public.ecr.aws/bottlerocket/bottlerocket-admin:v0.11.14";
```
needs to be v0.11.16

bcressey commented on sources/api/migration/migration-helpers/src/common_migrations.rs:
```diff
@@ -1912,3 +1788,250 @@ impl Migration for NoOpMigration {
+pub struct RemoveSchnauzerMigration {
+    pub setting: &'static str,
+    pub old_cmdline: &'static str,
+    pub new_cmdline: &'static str,
```
nit: unused?

bcressey reviewed (commented):
Nice!

cbgbt commented on Release.toml:
```diff
@@ -1,4 +1,8 @@
+<<<<<<< HEAD
 version = "1.32.0"
+=======
+version = "1.33.0"
+>>>>>>> dc5c6b00d (Upgrade Bottlerocket version to 1.33.0)
 
```
Looks like a merge conflict issue here.

cbgbt commented on sources/settings-migrations/v1.33.0/aws-remove-schnauzer-control/src/main.rs:
```diff
@@ -0,0 +1,27 @@
+    migrate(RemoveSchnauzerMigration {
+        setting: "settings.host-containers.control.source",
+        old_cmdline: OLD_CONTROL_CTR_CMDLINE,
+        new_cmdline: NEW_CONTROL_CTR_CMDLINE,
```
I think this value is unneeded?

cbgbt reviewed (commented):
nit: I'd spend a moment to clean up the commit messages if possible -- misspelling of `metadata` in a few commit messages titles, and the line wrapping is inconsistent in the GitHub UI for some of the commit message details (though this is sometimes just the GitHub UI being strange.).

cbgbt commented on sources/settings-migrations/v1.33.0/aws-remove-schnauzer-admin/src/main.rs:
```diff
@@ -0,0 +1,24 @@
+const OLD_ADMIN_CTR_CMDLINE: &str =
+    "schnauzer-v2 render --requires 'aws@v1(helpers=[ecr-prefix])' --template '{{ ecr-prefix settings.aws.region }}/bottlerocket-admin:v0.11.16'";
+
+/// We bumped the version of the default admin container
```
nit: inaccurate comment.

cbgbt commented on sources/settings-migrations/v1.33.0/aws-remove-schnauzer-control/src/main.rs:
```diff
@@ -0,0 +1,24 @@
+const OLD_CONTROL_CTR_CMDLINE: &str =
+    "schnauzer-v2 render --requires 'aws@v1(helpers=[ecr-prefix])' --template '{{ ecr-prefix settings.aws.region }}/bottlerocket-control:v0.7.20'";
+
+/// We bumped the version of the default control container
```
nit: comment needs updating.

bcressey reviewed (approved):
Looks good. Please double-check the commit message for this one as it refers to an earlier version of the changes:
```
common_migrations: add new migrations to handle settings-generator as struct
```

If we don't strictly need them, consider backing out the changes to `ReplaceSchnauzerMigration`, or else splitting them into a separate commit that explains why.

vyaghras commented:
☝️ Reverted the changes in ReplaceSchnauzerMigration and updated the commit.

cbgbt reviewed (approved):
I separately sent a few nits for the commit messages, but the code looks good to me. Nice work!

vyaghras committed:
migrations: remove metadata and data AWS Admin

In this migration we will remove the data and metadata if it matches
with existing ones in datastore.We will rely on storewolf and sundog
to populate correct metadata and data.

vyaghras committed:
migrations: remove metadata and data AWS control

In this migration we will remove setting-generator metadata and source
data if it matches with existing ones in datastore. We will rely on
storewolf and sundog to populate correct metadata and data.

vyaghras committed:
migrations: remove metadata and data public admin

In this migration we will remove the data if it matches
with existing ones in datastore. We will rely on storewolf
and sundog to populate correct data from defaults.

vyaghras committed:
migrations: remove metadata and data public control

In this migration we will remove the data if it matches
with existing ones in datastore. We will rely on
Storewolf and Sundog to populate correct data.

vyaghras committed:
migrations: remove all metadata and weak settings

On downgrade we will remove all the weak metadata and settings,
so that it can be populated from defaults.
